storage/innobase/log/log0recv.cc

```diff
@@ -2233,6 +2224,25 @@ ATTRIBUTE_COLD void recv_sys_t::wait_for_pool(size_t pages)
     buf_flush_sync_batch(recovered_lsn);
 }
 
+/** Check if a FIL_PAGE_LSN is valid during recovery.
+@param lsn    the FIL_PAGE_LSN
+@return the current log sequence number
+@retval 0     if the current log sequence number is unknown */
+ATTRIBUTE_COLD lsn_t recv_sys_t::check_page_lsn(lsn_t lsn)
```

Comment:
The return value 0 may mean two different things:
* The `FIL_PAGE_LSN` does not exceed the currently known LSN.
* Crash recovery is being executed in multiple batches, and we do not know yet whether the `FIL_PAGE_LSN` is in the future. It will be checked later based on `recv_sys.max_page_lsn`.

I will double-check this logic. It seems to me that before we can start any recovery (whether in a single batch or multiple batches), we must determine the final LSN. If this is the case, there should be no reason for `recv_sys.max_page_lsn` to exist.

Reply:
I replaced the function with a straightforward `void recv_sys_t::set_max_page_lsn(lsn_t lsn)` that will merely update `recv_sys.max_page_lsn`.

It turns out that the `log_sys.lsn` reflects the recovered LSN during this part of recovery (validating the doublewrite buffer or reading data pages). Therefore, we can simplify things quite a bit.

Reply:
3000bbd35d4955b2cfc0d7dfb2221234d894e126 does away with this function. The variables `recv_max_page_lsn` and `recv_lsn_checks_on` are being removed without any replacement. Whenever we read a page into the buffer pool, the `log_sys.lsn` must not be earlier than the `FIL_PAGE_LSN`. Only the page reads before the `buf_dblwr.recover()` invocation are skipping the LSN check.